pkg/cloudprovider/providers/azure/azure_loadbalancer.go

```diff
@@ -0,0 +1,623 @@
+				SourceAddressPrefix:      to.StringPtr("Internet"),
+				DestinationAddressPrefix: to.StringPtr("*"),
+				Access:    network.Allow,
+				Direction: network.Inbound,
+			},
+		}
+	}
+
+	// update security rules
+	dirtySg := false
+	updatedRules := []network.SecurityRule{}
+	if sg.Properties.SecurityRules != nil {
+		updatedRules = *sg.Properties.SecurityRules
+	}
+	// update security rules: remove unwanted
+	for i := len(updatedRules) - 1; i >= 0; i-- {
+		existingRule := updatedRules[i]
+		if serviceOwnsRule(service, *existingRule.Name) {
+			glog.Infof("reconcile(%s)(%t): sg rule(%s) - considering evicting", serviceName, wantLb, *existingRule.Name)
+			keepRule := false
+
+			for _, expectedRule := range expectedSecurityRules {
+				if strings.EqualFold(*existingRule.Name, *expectedRule.Name) {
+					glog.Infof("reconcile(%s)(%t): sg rule(%s) - keeping", serviceName, wantLb, *existingRule.Name)
+					keepRule = true
+					break
+				}
+			}
+			if !keepRule {
+				glog.Infof("reconcile(%s)(%t): sg rule(%s) - dropping", serviceName, wantLb, *existingRule.Name)
+				updatedRules = append(updatedRules[:i], updatedRules[i+1:]...)
+				dirtySg = true
+			}
+		}
+	}
+	// update security rules: add needed
+	for _, expectedRule := range expectedSecurityRules {
+		foundRule := false
+		for _, existingRule := range *sg.Properties.SecurityRules {
+			if strings.EqualFold(*existingRule.Name, *expectedRule.Name) {
+				glog.Infof("reconcile(%s)(%t): sg rule(%s) - already exists", serviceName, wantLb, *existingRule.Name)
+				foundRule = true
+				break
+			}
+		}
+		if !foundRule {
+			glog.Infof("reconcile(%s)(%t): sg rule(%s) - adding", serviceName, wantLb, *expectedRule.Name)
+
+			nextAvailablePriority, err := getNextAvailablePriority(updatedRules)
+			if err != nil {
+				return sg, false, err
+			}
+
+			expectedRule.Properties.Priority = to.Int32Ptr(nextAvailablePriority)
+			updatedRules = append(updatedRules, expectedRule)
+			dirtySg = true
+		}
+	}
+	if dirtySg {
+		glog.Infof("reconcile(%s)(%t): sg(%s) - dirty", serviceName, wantLb, az.SecurityGroupName)
+		sg.Properties.SecurityRules = &updatedRules
+	} else {
+		glog.Infof("reconcile(%s)(%t): sg(%s) - unchanged", serviceName, wantLb, az.SecurityGroupName)
+	}
+	return sg, dirtySg, nil
+}
+
+// This ensures the given VM's Primary NIC's Primary IP Configuration is
+// participating in the specified LoadBalancer Backend Pool.
+func (az *AzureCloud) ensureHostInPool(serviceName, machineName string, backendPoolID string) error {
+	glog.Infof("nicupdate(%s): vm(%s) - retrieving", serviceName, machineName)
+	machine, err := az.VirtualMachinesClient.Get(az.ResourceGroup, machineName, "")
+	if err != nil {
+		glog.Errorf("nicupdate(%s): vm(%s) - retrieving failed: %q", serviceName, machineName, err)
+		return fmt.Errorf("failed to retrieve vm instance. instance=%q", machineName)
+	}
+
+	primaryNicID, err := getPrimaryNicID(machine)
```

Comment:
`getPrimaryNICID` is accurate though, while I feel `getPrimaryIDForNIC` is not. (It's retrieving the NIC ID, and passing in the machine object, rather than getting the NIC object by it's ID as I feel `getPrimaryIDForNIC` would imply).

Maybe something like `getPrimaryNetworkInterfaceID` since the `azure-sdk-for-go` already refers to the NICs as "Interface".